File: javascript/src/xid_codec.js
/**
 * XID v1.1 二进制层编解码。
 */

import { OType } from './typed_value.js';

const MASK8 = 0xff;
const MASK16 = 0xffff;
const MASK32 = 0xffffffff;

function maskForBits(tbits) {
  if (tbits === 64) return -1; // 用 BigInt 路径处理
  if (tbits === 32) return MASK32;
  if (tbits === 16) return MASK16;
  if (tbits === 8) return MASK8;
  return (1 << tbits) - 1;
}

function pow2(n) {
  return n >= 32 ? 2 ** n : (1 << n);
}

const SW_BYTES = [1, 2, 4, 8];
const EMBEDDED_OTYPE = [
  [OType.UINT8, OType.UINT16, OType.UINT32, OType.UINT64],
  [OType.INT8, OType.INT16, OType.INT32, OType.INT64],
];

/** @param {import('./idmix.js').IdMix} m @param {import('./typed_value.js').TypedValue[]} typed @param {number} variantId */
export function encodeBinary(m, typed, variantId) {
  const parts = [];
  for (const tv of typed) parts.push(...encodeObject(tv));
  let objects = new Uint8Array(parts);
  const mask = (variantId * 0x9d + 0x37) & 0xff;
  objects = new Uint8Array(objects.map(b => b ^ mask));

  const count = typed.length;
  let header = (variantId << m.variantShift) | (count << m.countShift);
  const data = new Uint8Array(2 + objects.length);
  data[0] = header & 0xff;
  data[1] = (header >> 8) & 0xff;
  data.set(objects, 2);

  let xorSum = 0;
  for (const b of data) xorSum ^= b;
  header |= xorSum & m.checkMask;
  data[0] = header & 0xff;
  data[1] = (header >> 8) & 0xff;
  return data;
}

/** @param {import('./idmix.js').IdMix} m @param {Uint8Array} data */
export function decodeBinary(m, data) {
  if (data.length < 2) throw new Error('invalid data: too short');
  const header = data[0] | (data[1] << 8);
  const check = header & m.checkMask;
  const count = (header & m.countMask) >> m.countShift;
  const variantId = (header & m.variantMask) >> m.variantShift;

  if (variantId >= m.maxVariants) throw new Error(`invalid variant_id ${variantId}`);
  if (count > m.maxObjects) throw new Error(`invalid count ${count}`);

  const verify = new Uint8Array(data);
  verify[0] &= ~m.checkMask;
  let xorSum = 0;
  for (const b of verify) xorSum ^= b;
  if ((xorSum & m.checkMask) !== check) throw new Error('checksum mismatch');

  let objects = new Uint8Array(data.slice(2));
  const mask = (variantId * 0x9d + 0x37) & 0xff;
  objects = new Uint8Array(objects.map(b => b ^ mask));

  const result = [];
  let pos = 0;
  for (let i = 0; i < count; i++) {
    if (pos >= objects.length) throw new Error('premature end of data');
    const [tv, n] = decodeObject(objects.slice(pos));
    result.push(tv);
    pos += n;
  }
  if (pos !== objects.length) throw new Error('extra bytes after data objects');
  return result;
}

/** @param {import('./typed_value.js').TypedValue} tv */
function encodeObject(tv) {
  validateRange(tv.otype, tv.val);
  const embedded = tryEmbeddedHead(tv.otype, tv.val);
  if (embedded !== null) return [embedded & 0xff];
  const [mag, neg] = magnitudeFromTyped(tv.otype, tv.val);
  const sw = swFromMagnitude(mag);
  const payload = uintToLEBytes(mag, SW_BYTES[sw]);
  let head = 0x80 | (sw << 4) | tv.otype;
  if (neg) head |= 1 << 6;
  return [head, ...payload];
}

/** @param {Uint8Array} data */
function decodeObject(data) {
  if (!data.length) throw new Error('truncated object header');
  const head = data[0];
  if ((head & 0x80) === 0) {
    const sign = (head >> 6) & 1;
    const wb = (head >> 4) & 3;
    const v = head & 0x0f;
    const otype = EMBEDDED_OTYPE[sign][wb];
    const val = sign === 0 ? v : -v - 1;
    return [{ otype, val }, 1];
  }
  const sw = (head >> 4) & 3;
  const otype = head & 0x0f;
  if (otype > OType.INT64) throw new Error(`invalid otype ${otype}`);
  const numBytes = SW_BYTES[sw];
  if (data.length < 1 + numBytes) throw new Error('truncated object payload');
  let mag = 0n;
  for (let i = 0; i < numBytes; i++) mag |= BigInt(data[1 + i]) << BigInt(8 * i);
  const neg = ((head >> 6) & 1) !== 0;
  const val = valueFromMagnitude(mag, neg);
  validateRange(otype, val);
  return [{ otype, val }, 1 + numBytes];
}

function isUnsigned(otype) { return otype <= OType.UINT64; }
function isSigned(otype) { return otype >= OType.INT8; }

function widthBits(otype) {
  const m = { [OType.UINT8]: 0, [OType.INT8]: 0, [OType.UINT16]: 1, [OType.INT16]: 1,
              [OType.UINT32]: 2, [OType.INT32]: 2 };
  return m[otype] ?? 3;
}

function magnitudeFromTyped(otype, val) {
  const bval = typeof val === 'bigint' ? val : BigInt(val);
  if (isUnsigned(otype)) return [BigInt.asUintN(64, bval), false];
  if (bval < 0n) return [-bval, true];
  return [bval, false];
}

function swFromMagnitude(mag) {
  const m = typeof mag === 'bigint' ? mag : BigInt(mag);
  if (m < 256n) return 0;
  if (m < 65536n) return 1;
  if (m < 4294967296n) return 2;
  return 3;
}

function tryEmbeddedHead(otype, val) {
  const [mag, neg] = magnitudeFromTyped(otype, val);
  if (mag >= 17n) return null;
  const wb = widthBits(otype);
  if (mag === 16n) {
    if (neg) return (1 << 6) | (wb << 4) | 15;
    return null;
  }
  if (neg) return (1 << 6) | (wb << 4) | Number(mag - 1n);
  return (wb << 4) | Number(mag);
}

function valueFromMagnitude(mag, neg) {
  const m = typeof mag === 'bigint' ? mag : BigInt(mag);
  if (!neg) return m <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(m) : m;
  if (m === 1n << 63n) return -(1n << 63n);
  const v = -m;
  return v >= BigInt(Number.MIN_SAFE_INTEGER) && v <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(v) : v;
}

function uintToLEBytes(v, size) {
  const bval = typeof v === 'bigint' ? v : BigInt(v);
  const buf = [];
  for (let i = 0; i < size; i++) buf.push(Number((bval >> BigInt(8 * i)) & 0xffn));
  return buf;
}

function validateRange(otype, val) {
  const bval = typeof val === 'bigint' ? val : BigInt(val);
  const ranges = {
    [OType.UINT8]: [0n, 0xffn], [OType.UINT16]: [0n, 0xffffn], [OType.UINT32]: [0n, 0xffffffffn],
    [OType.UINT64]: [0n, 0xffffffffffffffffn], [OType.INT8]: [-128n, 127n], [OType.INT16]: [-32768n, 32767n],
    [OType.INT32]: [-2147483648n, 2147483647n], [OType.INT64]: [-0x8000000000000000n, 0x7fffffffffffffffn],
  };
  const [lo, hi] = ranges[otype] ?? [0n, 0n];
  if (bval < lo || bval > hi) throw new Error(`value ${val} out of range for otype ${otype}`);
}

File: python/idmix/xid_codec.py
"""XID v1.1 二进制层编解码。"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

from .typed_value import OType, TypedValue

if TYPE_CHECKING:
    from .idmix import IdMix

SW_BYTES = (1, 2, 4, 8)

EMBEDDED_OTYPE = (
    (OType.UINT8, OType.UINT16, OType.UINT32, OType.UINT64),
    (OType.INT8, OType.INT16, OType.INT32, OType.INT64),
)


def encode_binary(m: IdMix, typed: list[TypedValue], variant_id: int) -> bytes:
    """typedValue 序列 → XID 二进制块。"""
    objects = bytearray()
    for tv in typed:
        objects.extend(encode_object(tv))
    mask = (variant_id * 0x9D + 0x37) & 0xFF
    objects = bytes(b ^ mask for b in objects)

    count = len(typed)
    header = (variant_id << m.variant_shift) | (count << m.count_shift)
    data = bytearray(struct.pack("<H", header) + objects)

    xor_sum = 0
    for b in data:
        xor_sum ^= b
    check = xor_sum & m.check_mask
    header |= check
    data[0:2] = struct.pack("<H", header)
    return bytes(data)


def decode_binary(m: IdMix, data: bytes) -> list[TypedValue]:
    """XID 二进制块 → typedValue 序列。"""
    if len(data) < 2:
        raise ValueError("invalid data: too short")
    header = struct.unpack("<H", data[:2])[0]
    check = header & m.check_mask
    count = (header & m.count_mask) >> m.count_shift
    variant_id = (header & m.variant_mask) >> m.variant_shift

    if variant_id >= m.max_variants:
        raise ValueError(f"invalid variant_id {variant_id}")
    if count > m.max_objects:
        raise ValueError(f"invalid count {count}")

    verify = bytearray(data)
    verify[0] &= ~m.check_mask & 0xFF
    xor_sum = 0
    for b in verify:
        xor_sum ^= b
    if (xor_sum & m.check_mask) != check:
        raise ValueError("checksum mismatch")

    objects = bytearray(data[2:])
    mask = (variant_id * 0x9D + 0x37) & 0xFF
    objects = bytes(b ^ mask for b in objects)

    result: list[TypedValue] = []
    pos = 0
    for i in range(count):
        if pos >= len(objects):
            raise ValueError("premature end of data")
        tv, n = decode_object(objects[pos:])
        result.append(tv)
        pos += n
    if pos != len(objects):
        raise ValueError("extra bytes after data objects")
    return result


def encode_object(tv: TypedValue) -> bytes:
    validate_range(tv.otype, tv.val)
    if (head := _try_embedded_head(tv.otype, tv.val)) is not None:
        return bytes([head])
    mag, neg = _magnitude_from_typed(tv.otype, tv.val)
    sw = _sw_from_magnitude(mag)
    payload = _uint_to_le_bytes(mag, SW_BYTES[sw])
    head = 0x80 | (sw << 4) | tv.otype
    if neg:
        head |= 1 << 6
    return bytes([head]) + payload


def decode_object(data: bytes) -> tuple[TypedValue, int]:
    if not data:
        raise ValueError("truncated object header")
    head = data[0]
    if (head & 0x80) == 0:
        sign = (head >> 6) & 1
        wb = (head >> 4) & 0x03
        v = head & 0x0F
        otype = EMBEDDED_OTYPE[sign][wb]
        val = v if sign == 0 else -v - 1
        return TypedValue(otype, val), 1
    sw = (head >> 4) & 0x03
    otype = head & 0x0F
    if otype > OType.INT64:
        raise ValueError(f"invalid otype {otype}")
    num_bytes = SW_BYTES[sw]
    if len(data) < 1 + num_bytes:
        raise ValueError("truncated object payload")
    mag = 0
    for i in range(num_bytes):
        mag |= data[1 + i] << (8 * i)
    neg = ((head >> 6) & 1) != 0
    val = _value_from_magnitude(mag, neg)
    validate_range(otype, val)
    return TypedValue(otype, val), 1 + num_bytes


def _is_unsigned(otype: int) -> bool:
    return otype <= OType.UINT64


def _is_signed(otype: int) -> bool:
    return otype >= OType.INT8


def _width_bits(otype: int) -> int:
    return {OType.UINT8: 0, OType.INT8: 0, OType.UINT16: 1, OType.INT16: 1,
            OType.UINT32: 2, OType.INT32: 2}.get(otype, 3)


def _magnitude_from_typed(otype: int, val: int) -> tuple[int, bool]:
    if _is_unsigned(otype):
        return val & 0xFFFFFFFFFFFFFFFF, False
    if val < 0:
        return -val, True
    return val, False


def _sw_from_magnitude(mag: int) -> int:
    if mag < 256:
        return 0
    if mag < 65536:
        return 1
    if mag < 4294967296:
        return 2
    return 3


def _try_embedded_head(otype: int, val: int) -> int | None:
    mag, neg = _magnitude_from_typed(otype, val)
    if mag >= 17:
        return None
    wb = _width_bits(otype)
    if mag == 16:
        if neg:
            return (1 << 6) | (wb << 4) | 15
        return None
    if neg:
        return (1 << 6) | (wb << 4) | (mag - 1)
    return (wb << 4) | mag


def _value_from_magnitude(mag: int, neg: bool) -> int:
    if not neg:
        return mag
    if mag == 1 << 63:
        return -(1 << 63)
    return -mag


def _uint_to_le_bytes(v: int, size: int) -> bytes:
    return bytes((v >> (8 * i)) & 0xFF for i in range(size))


def _validate_range(otype: int, val: int) -> None:
    limits = {
        OType.UINT8: (0, 0xFF),
        OType.UINT16: (0, 0xFFFF),
        OType.UINT32: (0, 0xFFFFFFFF),
        OType.UINT64: (None, None),
        OType.INT8: (-128, 127),
        OType.INT16: (-32768, 32767),
        OType.INT32: (-2147483648, 2147483647),
        OType.INT64: (None, None),
    }
    lo, hi = limits.get(otype, (None, None))
    if lo is not None and val < lo:
        raise ValueError(f"value {val} out of range for otype {otype}")
    if hi is not None and val > hi:
        raise ValueError(f"value {val} out of range for otype {otype}")


def validate_range(otype: int, val: int) -> None:
    _validate_range(otype, val)
